### PR TITLE
Add Symlinks to FileSystem

### PR DESCRIPTION
**Pull request by tgymnich**
I've added symbolic links to `FileSystem` for use in: https://github.com/apple/swift-package-manager/pull/2985

### CODE REVIEW

File: Sources/TSCBasic/PathShims.swift
Comment (tgymnich) [inline Sources/TSCBasic/PathShims.swift:65]: should we keep `relative`? Or is this not important.

File: Tests/TSCBasicTests/FileSystemTests.swift
Comment (tgymnich) [inline Tests/TSCBasicTests/FileSystemTests.swift:57]: changed this for consistency

File: Sources/TSCBasic/FileSystem.swift
Comment (tgymnich) [inline Sources/TSCBasic/FileSystem.swift:954]: This seemed to be a bug

**Comment by MaxDesiatov** [comment]
@swift-ci please test

File: Sources/TSCBasic/FileSystem.swift
Comment (abertelrud) [reply Sources/TSCBasic/FileSystem.swift:954]: I agree, that seems to have a bug all along.  Good catch!

File: Sources/TSCBasic/FileSystem.swift
Comment (abertelrud) [inline Sources/TSCBasic/FileSystem.swift:353]: Commented on this below, but I think the relative form of this is still important.  The code from the original method:
```
    let destString = relative ? dest.relative(to: path.parentDirectory).pathString : dest.pathString
```
still seems relevant here.

File: Sources/TSCBasic/FileSystem.swift
Comment (abertelrud) [inline Sources/TSCBasic/FileSystem.swift:498]: I realize that supporting relative symlinks makes this a bit more tricky.  It might be best here to store a `String` with the actual symlink contents and then convert it to a path when used.  There should be an `AbsolutePath` initializer that takes a base path and a string representation of a path, and handles the case of both relative and an absolute string representation (using the base path only if the string representation is relative).

File: Sources/TSCBasic/FileSystem.swift
Comment (abertelrud) [inline Sources/TSCBasic/FileSystem.swift:580]: If the payload of the symlink is a string, this would then be something like `AbsolutePath(destination, relativeTo: path.parentDirectory)`.

File: Sources/TSCBasic/PathShims.swift
Comment (abertelrud) [reply Sources/TSCBasic/PathShims.swift:65]: The relative form is still important, since otherwise, only absolute-path symlinks could be created.

I'm not sure of the provenance of these freestanding functions and whether this should be deprecated or not.  Based on the `currentWorkingDirectory` property below it, it seems that the intent is to deprecate them and replace use with `localFileSystem`, which makes sense.

**Comment by tgymnich** [comment]
I went ahead and implemented relative symlinks.

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
Thanks, @tgymnich!  This fills in an important missing piece!  The changes look good to me; @neonichu is back this week and should take a look at the diffs as well, but from my point of view this looks ready.  Thanks!

**Review by abertelrud — APPROVED** [review]

**Comment by abertelrud** [comment]
Looks like a linux-specific crash in `FileSystemTests.testInMemoryFileSystemFileLock`:

```
Test Suite 'Selected tests' started at 2020-11-02 16:52:27.939
Test Suite 'FileSystemTests' started at 2020-11-02 16:52:27.941
Test Case 'FileSystemTests.testInMemoryFileSystemFileLock' started at 2020-11-02 16:52:27.942
Exited with signal code 11
```

**Review by neonichu — APPROVED** [review]

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by tgymnich** [comment]
At first I thought the issue on linux was related to multi-threaded dictionary access but it turns out that `Dictionary` segfaults on linux when accessing a key that does not exist.
 
In the tests, the file we want to lock and pass to `FileSystem.withLock<T>(on path: AbsolutePath, type: FileLock.LockType, _ body: () throws -> T) throws -> T` does not exist (which should be fine).
```
frame #10: 0x0000555555584ba8 Debugger`closure #1 in closure #1 in testInMemoryFileSystemFileLock(fs=0x0000555555897880, lockFile=TSCBasic.AbsolutePath @ 0x00007ffff1a34810, fileA=TSCBasic.AbsolutePath @ 0x00007ffff1a34800, fileB=TSCBasic.AbsolutePath @ 0x00007ffff1a347f0) at main.swift:15:25
   12  	
   13  	        let writerThreads = (0..<100).map { _ in
   14  	            return Thread {
-> 15  	                try! fs.withLock(on: lockFile, type: .exclusive) {   // lockFile = <AbsolutePath:"/lockfile">
   16  	                    // Get thr current contents of the file if any.
   17  	                    let valueA: Int
   18  	                    if fs.exists(fileA) {
```

This should work the same as for `FileLock`, which creates this lockfile for us when using `LocalFileSystem`. However when we use `InMemoryFileSystem` we try to get the node at the given path to resolve the symlink, in case it is s a symlink if it's not we just use the path as is.
```
frame #4: 0x00005555555ba3f3 Debugger`closure #1 in InMemoryFileSystem.withLock<A>(self=0x0000555555897880, path=TSCBasic.AbsolutePath @ 0x00007ffff1a343f0) at FileSystem.swift:849:53
   846 	            let resolvedPath: AbsolutePath
   847 	
   848 	            // FIXME: resolving symlinks is not yet thread safe
-> 849 	            if case let .symlink(destination) = try getNode(path)?.contents {   // path = <AbsolutePath:"/lockfile">
   850 	                resolvedPath = AbsolutePath(destination, relativeTo: path.parentDirectory)
   851 	            } else {
   852 	                resolvedPath = path
```

On linux `getNode` causes a segfault for some reason when trying to get the node (which does not exist as expected).
```
frame #2: 0x00005555555b6127 Debugger`getNodeInternal #1 (path=TSCBasic.AbsolutePath @ 0x00007ffff1a340b8, self=0x0000555555897880, followSymlink=true) in InMemoryFileSystem.getNode(_:followSymlink:) at FileSystem.swift:577:40
   574 	            }
   575 	
   576 	            // Return the directory entry.
-> 577 	            let node = contents.entries[path.basename]   // path.basename = "/lockfile"
   578 	
   579 	            switch node?.contents {
   580 	            case .directory, .file:
```

`contents.entries` and file system dump for reference.
```
(lldb) expr self.dumpTree()
.
├── fileB
└── fileA
() $R22 = {}
(lldb) po contents.entries
▿ 2 elements
  ▿ 0 : 2 elements
    - key : "fileB"
    ▿ value : <Node: 0x7fffe8000cb0>
  ▿ 1 : 2 elements
    - key : "fileA"
    ▿ value : <Node: 0x7fffe4000d40>
```

whole backtrace for reference
```
* thread #2, name = 'Debugger', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
    frame #0: 0x00007ffff7bb5757 libswiftCore.so`Swift.Dictionary._Variant.lookup(A) -> Swift.Optional<B> + 23
    frame #1: 0x00007ffff7df8ee9 libswiftCore.so`Swift.Dictionary.subscript.getter : (A) -> Swift.Optional<B> + 9
    frame #2: 0x00005555555b6127 Debugger`getNodeInternal #1 (path=TSCBasic.AbsolutePath @ 0x00007ffff1a340b8, self=0x0000555555897880, followSymlink=true) in InMemoryFileSystem.getNode(_:followSymlink:) at FileSystem.swift:577:40
    frame #3: 0x00005555555b5d0b Debugger`InMemoryFileSystem.getNode(path=TSCBasic.AbsolutePath @ 0x00007ffff1a34110, followSymlink=true, self=0x0000555555897880) at FileSystem.swift:591:20
    frame #4: 0x00005555555ba3f3 Debugger`closure #1 in InMemoryFileSystem.withLock<A>(self=0x0000555555897880, path=TSCBasic.AbsolutePath @ 0x00007ffff1a343f0) at FileSystem.swift:849:53
    frame #5: 0x00005555555baa38 Debugger`partial apply for closure #1 in InMemoryFileSystem.withLock<A>(on:type:_:) at <compiler-generated>:0
    frame #6: 0x00005555555baa93 Debugger`thunk for @callee_guaranteed () -> (@owned DispatchQueue, @error @owned Error) at <compiler-generated>:0
    frame #7: 0x00005555555baaf4 Debugger`partial apply for thunk for @callee_guaranteed () -> (@owned DispatchQueue, @error @owned Error) at <compiler-generated>:0
    frame #8: 0x00005555555dcac7 Debugger`Lock.withLock<T>(body=0x00005555555baae0 Debugger`partial apply forwarder for reabstraction thunk helper from @callee_guaranteed () -> (@owned Dispatch.DispatchQueue, @error @owned Swift.Error) to @escaping @callee_guaranteed () -> (@out Dispatch.DispatchQueue, @error @owned Swift.Error) at <compiler-generated>, self=TSCBasic.Lock @ 0x00007ffff1a344f8) at Lock.swift:34:20
    frame #9: 0x00005555555ba159 Debugger`InMemoryFileSystem.withLock<T>(path=TSCBasic.AbsolutePath @ 0x00007ffff1a346c0, type=exclusive, body=0x0000555555585b00 Debugger`reabstraction thunk helper from @callee_guaranteed () -> (@error @owned Swift.Error) to @escaping @callee_guaranteed () -> (@out (), @error @owned Swift.Error)partial apply forwarder with unmangled suffix ".8" at <compiler-generated>, self=0x0000555555897880) at FileSystem.swift:844:58
  * frame #10: 0x0000555555584ba8 Debugger`closure #1 in closure #1 in testInMemoryFileSystemFileLock(fs=0x0000555555897880, lockFile=TSCBasic.AbsolutePath @ 0x00007ffff1a34810, fileA=TSCBasic.AbsolutePath @ 0x00007ffff1a34800, fileB=TSCBasic.AbsolutePath @ 0x00007ffff1a347f0) at main.swift:15:25
    frame #11: 0x0000555555585abf Debugger`partial apply for closure #1 in closure #1 in testInMemoryFileSystemFileLock() at <compiler-generated>:0
    frame #12: 0x0000555555619e3a Debugger`closure #1 in closure #1 in Thread.init(task=0x0000555555585a90 Debugger`partial apply forwarder for closure #1 () -> () in closure #1 (Swift.Int) -> TSCBasic.Thread in Debugger.testInMemoryFileSystemFileLock() throws -> () at <compiler-generated>, strongSelf=0x00005555558991e0) at Thread.swift:43:21
    frame #13: 0x00005555555852ff Debugger`thunk for @callee_guaranteed () -> (@error @owned Error) at <compiler-generated>:0
    frame #14: 0x000055555561a3d4 Debugger`thunk for @callee_guaranteed () -> (@error @owned Error)partial apply at <compiler-generated>:0
    frame #15: 0x000055555559e9a7 Debugger`Condition.whileLocked<T>(body=0x000055555561a3c0 Debugger`reabstraction thunk helper from @callee_guaranteed () -> (@error @owned Swift.Error) to @escaping @callee_guaranteed () -> (@out (), @error @owned Swift.Error)partial apply forwarder with unmangled suffix ".4" at <compiler-generated>, self=TSCBasic.Condition @ 0x00007ffff1a349d8) at Condition.swift:50:20
    frame #16: 0x0000555555619cd6 Debugger`closure #1 in Thread.init(self=0x00005555558991e0, task=0x0000555555585a90 Debugger`partial apply forwarder for closure #1 () -> () in closure #1 (Swift.Int) -> TSCBasic.Thread in Debugger.testInMemoryFileSystemFileLock() throws -> () at <compiler-generated>) at Thread.swift:42:46
    frame #17: 0x00007ffff755e086 libFoundation.so`Foundation.Thread.main() -> () + 22
    frame #18: 0x00007ffff755ce7f libFoundation.so`Foundation.(NSThreadStart in _643EAFDBA6441671515D9121EE76E49E)(Swift.Optional<Swift.UnsafeMutableRawPointer>) -> Swift.Optional<Swift.UnsafeMutableRawPointer> + 239
    frame #19: 0x00007ffff755e049 libFoundation.so`@objc Foundation.(NSThreadStart in _643EAFDBA6441671515D9121EE76E49E)(Swift.Optional<Swift.UnsafeMutableRawPointer>) -> Swift.Optional<Swift.UnsafeMutableRawPointer> + 9
    frame #20: 0x00007ffff7f9d609 libpthread.so.0`start_thread + 217
    frame #21: 0x00007ffff6cdd293 libc.so.6`clone + 67
```

I've created a swift package for debugging this exact issue: https://github.com/tgymnich/Debugger

**Comment by abertelrud** [comment]
Thank you very much @tgymnich for creating the separate test package.  I will try to make time in the next day or two to take a look, and meanwhile I am wondering if we should also involve someone working on Foundation would also have an idea.  I don't know who would be most appropriate but maybe @neonichu does.

**Comment by tgymnich** [comment]
I took another look at the issue using TSan. It looks like this is some kind of access race and likely not caused by Foundation but by the test itself.
```
==================
WARNING: ThreadSanitizer: Swift access race (pid=29584)
  Modifying access of Swift variable at 0x7b08000297f0 by thread T2 (mutexes: write M208):
    #0 InMemoryFileSystem.writeFileContents(_:bytes:) <null>:2 (Debugger:x86_64+0x1000548ed)
    #1 closure #1 in closure #1 in closure #1 in testInMemoryFileSystemFileLock() main.swift:25 (Debugger:x86_64+0x100004b81)
    #2 partial apply for closure #1 in closure #1 in closure #1 in testInMemoryFileSystemFileLock() <compiler-generated> (Debugger:x86_64+0x10000629b)
    #3 thunk for @callee_guaranteed () -> (@error @owned Error) <compiler-generated> (Debugger:x86_64+0x10000505f)
    #4 thunk for @callee_guaranteed () -> (@error @owned Error)partial apply <compiler-generated> (Debugger:x86_64+0x100006317)
    #5 partial apply for thunk for @callee_guaranteed () -> (@out A, @error @owned Error) <null>:2 (libswiftDispatch.dylib:x86_64+0x11250)
    #6 _dispatch_client_callout <null>:2 (libdispatch.dylib:x86_64+0x2825)
    #7 partial apply for OS_dispatch_queue._syncBarrier(block:) <null>:2 (libswiftDispatch.dylib:x86_64+0xe367)
    #8 closure #1 in closure #1 in testInMemoryFileSystemFileLock() main.swift:16 (Debugger:x86_64+0x10000463b)
    #9 partial apply for closure #1 in closure #1 in testInMemoryFileSystemFileLock() <compiler-generated> (Debugger:x86_64+0x10000622f)
    #10 closure #1 in closure #1 in Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f3970)
    #11 partial apply for closure #1 in closure #1 in Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f68d9)
    #12 thunk for @callee_guaranteed () -> (@error @owned Error) <null>:2 (Debugger:x86_64+0x10000738f)
    #13 thunk for @callee_guaranteed () -> (@error @owned Error)partial apply <null>:2 (Debugger:x86_64+0x1000f6937)
    #14 Condition.whileLocked<A>(_:) <null>:2 (Debugger:x86_64+0x100028644)
    #15 closure #1 in Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f3712)
    #16 partial apply for closure #1 in Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f3899)
    #17 thunk for @escaping @callee_guaranteed () -> () <null>:2 (Debugger:x86_64+0x1000f4423)
    #18 thunk for @escaping @callee_unowned @convention(block) () -> () <null>:2 (Debugger:x86_64+0x1000f46d4)
    #19 partial apply for thunk for @escaping @callee_unowned @convention(block) () -> () <null>:2 (Debugger:x86_64+0x1000f67f1)
    #20 ThreadImpl.main() <null>:2 (Debugger:x86_64+0x1000f42f9)
    #21 @objc ThreadImpl.main() <null>:2 (Debugger:x86_64+0x1000f4384)
    #22 __NSThread__start__ <null>:2 (Foundation:x86_64+0x587a1)

  Previous read of size 8 at 0x7b08000297f0 by thread T3 (mutexes: write M209, write M138):
    #0 getNodeInternal #1 (_:) in InMemoryFileSystem.getNode(_:followSymlink:) <null>:2 (Debugger:x86_64+0x100050440)
    #1 InMemoryFileSystem.getNode(_:followSymlink:) <null>:2 (Debugger:x86_64+0x10004fedd)
    #2 closure #1 in InMemoryFileSystem.withLock<A>(on:type:_:) <null>:2 (Debugger:x86_64+0x100056ca3)
    #3 partial apply for closure #1 in InMemoryFileSystem.withLock<A>(on:type:_:) <null>:2 (Debugger:x86_64+0x100057513)
    #4 thunk for @callee_guaranteed () -> (@owned OS_dispatch_queue, @error @owned Error) <null>:2 (Debugger:x86_64+0x100057593)
    #5 partial apply for thunk for @callee_guaranteed () -> (@owned OS_dispatch_queue, @error @owned Error) <null>:2 (Debugger:x86_64+0x100057667)
    #6 Lock.withLock<A>(_:) <null>:2 (Debugger:x86_64+0x10009486f)
    #7 InMemoryFileSystem.withLock<A>(on:type:_:) <null>:2 (Debugger:x86_64+0x1000566c5)
    #8 closure #1 in closure #1 in testInMemoryFileSystemFileLock() main.swift:16 (Debugger:x86_64+0x10000463b)
    #9 partial apply for closure #1 in closure #1 in testInMemoryFileSystemFileLock() <compiler-generated> (Debugger:x86_64+0x10000622f)
    #10 closure #1 in closure #1 in Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f3970)
    #11 partial apply for closure #1 in closure #1 in Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f68d9)
    #12 thunk for @callee_guaranteed () -> (@error @owned Error) <null>:2 (Debugger:x86_64+0x10000738f)
    #13 thunk for @callee_guaranteed () -> (@error @owned Error)partial apply <null>:2 (Debugger:x86_64+0x1000f6937)
    #14 Condition.whileLocked<A>(_:) <null>:2 (Debugger:x86_64+0x100028644)
    #15 closure #1 in Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f3712)
    #16 partial apply for closure #1 in Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f3899)
    #17 thunk for @escaping @callee_guaranteed () -> () <null>:2 (Debugger:x86_64+0x1000f4423)
    #18 thunk for @escaping @callee_unowned @convention(block) () -> () <null>:2 (Debugger:x86_64+0x1000f46d4)
    #19 partial apply for thunk for @escaping @callee_unowned @convention(block) () -> () <null>:2 (Debugger:x86_64+0x1000f67f1)
    #20 ThreadImpl.main() <null>:2 (Debugger:x86_64+0x1000f42f9)
    #21 @objc ThreadImpl.main() <null>:2 (Debugger:x86_64+0x1000f4384)
    #22 __NSThread__start__ <null>:2 (Foundation:x86_64+0x587a1)

  Location is heap block of size 24 at 0x7b08000297e0 allocated by main thread:
    #0 malloc <null>:3 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x5239a)
    #1 swift_slowAlloc <null>:2 (libswiftCore.dylib:x86_64+0x2f2ca8)
    #2 InMemoryFileSystem.init() <null>:2 (Debugger:x86_64+0x10004fac5)
    #3 InMemoryFileSystem.__allocating_init() <null>:2 (Debugger:x86_64+0x10004f951)
    #4 testInMemoryFileSystemFileLock() main.swift:5 (Debugger:x86_64+0x10000306e)
    #5 main main.swift:59 (Debugger:x86_64+0x100002eac)

  Mutex M208 (0x7b2400001c30) created at:
    #0 pthread_mutex_init <null>:3 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2e913)
    #1 -[NSCondition init] <null>:2 (Foundation:x86_64+0x2f01c)
    #2 NSCondition.__allocating_init() <null>:2 (Debugger:x86_64+0x100027a3f)
    #3 Condition.init() <null>:2 (Debugger:x86_64+0x100027b11)
    #4 Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f2fab)
    #5 Thread.__allocating_init(task:) <null>:2 (Debugger:x86_64+0x1000f2df4)
    #6 closure #1 in testInMemoryFileSystemFileLock() main.swift:15 (Debugger:x86_64+0x1000042e4)
    #7 partial apply for closure #1 in testInMemoryFileSystemFileLock() <compiler-generated> (Debugger:x86_64+0x10000437e)
    #8 thunk for @callee_guaranteed (@unowned Int) -> (@owned Thread, @error @owned Error) <compiler-generated> (Debugger:x86_64+0x10000511a)
    #9 partial apply for thunk for @callee_guaranteed (@unowned Int) -> (@owned Thread, @error @owned Error) <compiler-generated> (Debugger:x86_64+0x100005202)
    #10 Collection.map<A>(_:) <null>:2 (libswiftCore.dylib:x86_64+0x12504)
    #11 main main.swift:59 (Debugger:x86_64+0x100002eac)

  Mutex M209 (0x7b2400001e70) created at:
    #0 pthread_mutex_init <null>:3 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2e913)
    #1 -[NSCondition init] <null>:2 (Foundation:x86_64+0x2f01c)
    #2 NSCondition.__allocating_init() <null>:2 (Debugger:x86_64+0x100027a3f)
    #3 Condition.init() <null>:2 (Debugger:x86_64+0x100027b11)
    #4 Thread.init(task:) <null>:2 (Debugger:x86_64+0x1000f2fab)
    #5 Thread.__allocating_init(task:) <null>:2 (Debugger:x86_64+0x1000f2df4)
    #6 closure #1 in testInMemoryFileSystemFileLock() main.swift:15 (Debugger:x86_64+0x1000042e4)
    #7 partial apply for closure #1 in testInMemoryFileSystemFileLock() <compiler-generated> (Debugger:x86_64+0x10000437e)
    #8 thunk for @callee_guaranteed (@unowned Int) -> (@owned Thread, @error @owned Error) <compiler-generated> (Debugger:x86_64+0x10000511a)
    #9 partial apply for thunk for @callee_guaranteed (@unowned Int) -> (@owned Thread, @error @owned Error) <compiler-generated> (Debugger:x86_64+0x100005202)
    #10 Collection.map<A>(_:) <null>:2 (libswiftCore.dylib:x86_64+0x12504)
    #11 main main.swift:59 (Debugger:x86_64+0x100002eac)

  Mutex M138 (0x7b1800004030) created at:
    #0 pthread_mutex_init <null>:3 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2e913)
    #1 -[NSLock init] <null>:2 (Foundation:x86_64+0x3c3b)
    #2 NSLock.__allocating_init() <null>:2 (Debugger:x86_64+0x10009454f)
    #3 Lock.init() <null>:2 (Debugger:x86_64+0x100094621)
    #4 InMemoryFileSystem.init() <null>:2 (Debugger:x86_64+0x10004fa47)
    #5 InMemoryFileSystem.__allocating_init() <null>:2 (Debugger:x86_64+0x10004f951)
    #6 testInMemoryFileSystemFileLock() main.swift:5 (Debugger:x86_64+0x10000306e)
    #7 main main.swift:59 (Debugger:x86_64+0x100002eac)

  Thread T2 (tid=465955, running) created by main thread at:
    #0 pthread_create <null>:3 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2d8fd)
    #1 -[NSThread start] <null>:2 (Foundation:x86_64+0x582f1)
    #2 closure #3 in testInMemoryFileSystemFileLock() main.swift:52 (Debugger:x86_64+0x1000059ed)
    #3 thunk for @callee_guaranteed (@guaranteed Thread) -> (@error @owned Error) <compiler-generated> (Debugger:x86_64+0x100005a63)
    #4 partial apply for thunk for @callee_guaranteed (@guaranteed Thread) -> (@error @owned Error) <compiler-generated> (Debugger:x86_64+0x100005b1a)
    #5 Sequence.forEach(_:) <null>:2 (libswiftCore.dylib:x86_64+0x12064c)
    #6 main main.swift:59 (Debugger:x86_64+0x100002eac)

  Thread T3 (tid=465956, running) created by main thread at:
    #0 pthread_create <null>:3 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2d8fd)
    #1 -[NSThread start] <null>:2 (Foundation:x86_64+0x582f1)
    #2 closure #3 in testInMemoryFileSystemFileLock() main.swift:52 (Debugger:x86_64+0x1000059ed)
    #3 thunk for @callee_guaranteed (@guaranteed Thread) -> (@error @owned Error) <compiler-generated> (Debugger:x86_64+0x100005a63)
    #4 partial apply for thunk for @callee_guaranteed (@guaranteed Thread) -> (@error @owned Error) <compiler-generated> (Debugger:x86_64+0x100005b1a)
    #5 Sequence.forEach(_:) <null>:2 (libswiftCore.dylib:x86_64+0x12064c)
    #6 main main.swift:59 (Debugger:x86_64+0x100002eac)

SUMMARY: ThreadSanitizer: Swift access race (Debugger:x86_64+0x1000548ed) in InMemoryFileSystem.writeFileContents(_:bytes:)+0x81d
==================
```

**Comment by abertelrud** [comment]
That makes sense.  Hmm, the InMemoryFileSystem class doesn't look thread-safe at all...

**Comment by tgymnich** [comment]
I'm quite sure the file write in the test is racing the symlink resolution. I'll think of a way we can avoid this. Making `InMemoryFileSystem` thread safe would be out of scope for this PR.

**Comment by tgymnich** [comment]
I can see 3 possible solutions:

1.  Make `InMemoryFileSystem` thread-safe -> **out of scope for this PR**
2.  Protect every access of the `InMemoryFileSystem` with just one lock and get rid of per file locks. -> **bad for performance but otherwise ok**
3. Don't resolve relative symlinks. -> **functionally insufficient**

Maybe we should also question our reasoning for implementing (relative)symlinks at this stage.

Overall I think solution 2 would be best and I also think it's worth pursuing symlinks at this time since they will be useful beyond linking the cache directory into `~/.swiftpm/cache`.

**Comment by abertelrud** [comment]
I agree with the analysis here, and I think that #2 (a single lock in the memory file system) would actually be best:  the main (or even only, in practice) reason for InMemoryFileSystem is for unit tests, and I think that the semantics are much more important than the performance.  So I don't think, for example, that we should back off on a semantically correct symlink implementation.

Thanks for your work on this!

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
Thank you @tgymnich!  I know this whole effort has been a lot of work.  Do you feel that this is ready to merge once the tests pass?

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please test

**Pull request merged**
